docs/design/third-party-libraries.md

```diff
@@ -0,0 +1,175 @@
+this install fails, ignore the failure.
+
+  These packages can be used to wrap libraries that we want to support but can't update
+easily, like Jasmine or LESS.
+1. **Run `package["angular-cli"].scripts["postinstall"]` if it exists.**
+
+# Hooks
+The third party library can implement hooks into the scaffolding, and the build system. The
+CLI's tasks will look for the proper hooks prior to running and execute them.
+
+The order of execution of these hooks is breadth first, going through all node packages and
+checking for the `package['angular-cli']['hooks']['${hookName}']`. The hooks are then
+`require()`'d as is, from within the app root folder. Within the same level of the dependency
+tree, there is no guarantee for the order of execution.
+
+## Install Hooks
+The only tricky part here is install hooks. The installed package should recursively call
+its `(pre|post|)install` hooks only on packages that are newly installed. It should call
+`(pre|post|)reinstall` on those who were already installed. A map of the currently installed
+packages should be kept before performing `npm install`.
+
+
+# appData
+The `angular-cli` key in the generated app should be used for `angular-cli` specific data.
+This includes the CLI configuration itself, as well as third-parties library configuration.
+
+Third-parties can store data that will be passed to the app, and can use that data themselves.
+That data can be anything. Any keys starting with either `$` or `_` will be ignored and not
+passed to the client. It could be used for builds or scaffolding information.
+
+During installation, there's a step to prompt the user for data. The schema contains a prompt
+text and a default value. The default value type is used to convert the string entered by the
+user. The key used in `appPrompt` is the key saved in appData.
+
+Example:
+
+```typescript
+{ // ...
+  "angular-cli": {
+    "appPrompt": {
+      "number": {
+        "prompt": "Please enter a number:",
+        "defaultValue": 0
+      },
+      "url": {
+        "prompt": "URL of your website:",
+        "defaultValue": "${homepage}"
+      },
+    }
+  }
+}
+```
+
+The default value is necessary as a `quiet` mode is enforced, using default values to fill
+in the data without user interaction. The special strings `${...}` can be used to replace
+with special values in the default string values. The values are taken from the `package.json`.
+
+We use a declarative style to enforce sensible defaults and make sure developers think about
+this. *In the case where developers want more complex interaction, they can use the
+install/uninstall hooks to prompt users.* But 3rd party libraries developers need to be aware
+that those hooks might not prompt users (no STDIN) and throw an error.
+
+# Providers
+Adding Angular providers to the app should be seamless. The build process will create a
+`providers.js` from all the providers contained in all the dependencies. The User can blacklist
```

Comment:
If the build process will generate this file, it will be hard for the user to do anything with it as his changes will be overwritten.

If it's intended to be a sort of configuration file, it would probably be better to keep it as a json file that is processed... except that I don't think we can do dynamic imports in TS (nor ES2015).

Manipulating this file will pose similar problems to the ones our current `ng install` has.